[RFC] ci: install node host in Appveyor and Travis

janlazo opened the pull request:
Cherry-pick from #7412
The patch installs node client in Travis and Appveyor and currently tests if neovim can detect it.

~~Appveyor is not setup for testing in https://github.com/neovim/node-client so~~ I will test the node host ~~on Travis only and~~ with plain javascript (no typescript, babel, webpack, ES7). ES6 should be enough for basic tests.

janlazo committed:
ci: install node host in Appveyor and Travis

janlazo commented:
The following works in Windows. Perhaps plugin registration is the issue for node provider (or msgpack?).

hello.js

```javascript
const socket = process.env.NVIM_LISTEN_ADDRESS;
const nvim = require('neovim');
const host = nvim.attach({socket: socket});
host.command('echo "hello"');
// Close the socket and the terminate node 
process.exit(0);
```

then run it in nvim

``` 
call jobstart(['node', 'hello.js'])
```

`hello` should appear at the bottom of the screen.

justinmk commented:
Confused by your last comment. It looks like appveyor "ran" the `node_spec.lua` successfully:

```
Running tests from C:/projects/neovim/test/functional\provider\node_spec.lua
[----------] 0 tests from C:/projects/neovim/test/functional\provider\node_spec.lua (31.00 ms total)
```

It didn't print the warning message, so "0 tests" means there was no problem.

justinmk commented on test/functional/helpers.lua:
```diff
@@ -614,7 +614,7 @@ local function new_pipename()
 
 local function missing_provider(provider)
-  if provider == 'ruby' then
+  if provider == 'ruby' or provider == 'node' then
```
Just a general note: the nodejs host is not really a "provider", it's just a "host". I missed this when I merged that PR. 

Providers are shims for features that otherwise would be implemented in C (such as `:ruby`, `:python`, clipboard).

janlazo commented:
There are no `describe()` and `it()` after the `before_each` in node_spec.lua
I will add one later using `jobstart` based on https://github.com/neovim/node-client#nvim_listen_address

janlazo committed:
test: add test case for node client api

justinmk commented on test/functional/provider/node_spec.lua:
```diff
@@ -15,3 +19,22 @@ end
   clear()
 end)
+
+describe('node host', function()
```
Let's use "nodejs" everywhere possible. Reduces ambiguity when grepping, etc.

janlazo replied:
"node" is used in https://github.com/neovim/neovim/blob/master/runtime/autoload/provider/node.vim and `rplugin/node`.

justinmk replied:
Yes. I just mean all of the new instances in this PR. I'll change the other instances later, except for anything that might break clients.

janlazo committed:
ci: install nodejs 8 in Appveyor

janlazo committed:
test: rename node to nodejs

janlazo committed:
ci: set NODE_PATH for global require() in nodejs

janlazo committed:
ci: set NODE_PATH to nodejs default global path

janlazo committed:
ci: set NODE_PATH with `npm root -g`

justinmk commented on ci/run_tests.sh:
```diff
@@ -23,7 +23,7 @@ if test "$CLANG_SANITIZER" != "TSAN" ; then
   # doesn't happen in the unit/functional tests
   run_test run_unittests
-  NODE_PATH=/usr/local/lib/node_modules
+  NODE_PATH=$(npm root -g)
```
always a good idea to surround with quotes,

    NODE_PATH="$(npm root -g)"

janlazo committed:
ci: escape and export NODE_PATH on Travis

janlazo commented:
the test passes in Travis (node 6) and Appveyor (node 8).
Next is the plugin API.
Last is plugin registration which doesn't work in Windows.

janlazo commented:
I'm thinking of bundling in this PR the version check mentioned in https://github.com/neovim/neovim/pull/7577#issuecomment-350590592. Could be a useful warning in `:checkhealth` by requiring nodejs 6 at least.

It's unrelated to tests in `nodejs_spec.lua` but it should restore debug behaviour mentioned in the nodejs client repo.

janlazo commented:
Plugin registration didn't work because of slash normalization (https://github.com/neovim/neovim/issues/4795, not https://github.com/neovim/neovim/pull/7525). nodejs uses supports backslash as path separator on Windows but the paths its receiving use forward slash as path separator.
I can register plugin commands now (based on rplugin.vim) but the registered commands aren't working for some reason.

If only there's a way to force-load a nodejs plugin directly through the nodejs host ...

janlazo committed:
host.vim: don't normalize slashes for nodejs rplugins

https://github.com/neovim/neovim/commit/46a080015ff83aaf74e26b4517a6620815cfa078
prevents plugin registration on Windows.
nodejs supports backslash as path separator on Windows, not forward slash.

janlazo commented:
@billyvg How can I load plugin directly without creating `rplugin/node/my_plugin`? I tried creating a new `Host` instance and use `getPlugin` but it's not working for me. I tried passing `nvim`, attached through the socket, to `Host` but did nothing.

janlazo committed:
provider: check node version for debug support

Resolve https://github.com/neovim/neovim/pull/7577#issuecomment-350590592 for Unix.

janlazo committed:
provider: test if nodejs in ci supports --inspect-brk

janlazo committed:
test: fix missing comma

janlazo committed:
health: add node host warnings

nodejs host for neovim requires nodejs 6+ to work properly.
nodejs 6.12+ or 7.6+ is required for debug support via `node --inspect-brk`.

janlazo committed:
test: add one for nodejs host plugin API

janlazo committed:
health: $NVIM_NODE_HOST_DEBUG is no-opt on Windows

janlazo commented:
I can add one more test for `rplugin/node` but I can't confirm that it will work on Windows  so I'll stop here for now.

justinmk commented on appveyor.yml:
```diff
@@ -9,6 +9,8 @@ matrix:
     - configuration: MINGW_64-gcov
 install: []
+before_build:
+- ps: Install-Product node 8
```
This takes 20s according to appveyor logs. But it doesn't mention a path, so we can't add it to cache...

justinmk commented:
> I'd say that the workaround should come from nodejs host directly instead of bypassing the shims, generated by npm.

Can you create a request at https://github.com/neovim/node-client/

justinmk commented on ci/run_tests.sh:
```diff
@@ -23,6 +23,7 @@ if test "$CLANG_SANITIZER" != "TSAN" ; then
   # Additional threads are only created when the builtin UI starts, which
   # doesn't happen in the unit/functional tests
   run_test run_unittests
+  export NODE_PATH="$(npm root -g)"
```
Doing this here is ugly, but I don't see a better way.

janlazo replied:
I think plugin authors are expected to locally install the neovim npm package as a dependency in `package.json` to avoid global `require()`. The test cases I've created use the API directly instead of the relying on the host ([bin/cli.js](https://github.com/neovim/node-client/blob/master/bin/cli.js)) so `NODE_PATH` must be set.

I wonder if there's a way to break up the npm package into host-only (end users, global install) and plugin-only (plugin authors, local install) without source code duplication.

billyvg replied:
I've been working on a branch that does this (host, plugin, "attach"/api), just have been really busy lately.

justinmk commented on runtime/autoload/health/provider.vim:
```diff
@@ -502,7 +502,17 @@ function! s:check_node() abort
+    endif
+  endif
+  if has('win32') || !provider#node#can_inspect()
+    call health#report_warn('$NVIM_NODE_HOST_DEBUG is no-op.')
```
Users will be confused by this. How about:

    The nodejs version on this system does not support --inspect-brk, $NVIM_NODE_HOST_DEBUG will be ignored.

justinmk commented on runtime/autoload/health/provider.vim:
```diff
@@ -502,7 +502,17 @@ function! s:check_node() abort
+  call health#report_info('Node: '. node_v)
+  if !s:shell_error
+    let node_v = split(split(node_v, "\n")[0][1:], '\.')
+    if node_v[0] < 6
```
Does `s:version_cmp()` work?

justinmk commented on runtime/autoload/health/provider.vim:
```diff
@@ -502,7 +502,17 @@ function! s:check_node() abort
+  if !s:shell_error
+    let node_v = split(split(node_v, "\n")[0][1:], '\.')
+    if node_v[0] < 6
+      call health#report_warn('node host does not support your node installation.')
```
'node host does not support node version '.node_v

justinmk commented on runtime/autoload/provider/node.vim:
```diff
@@ -24,7 +44,7 @@ function! provider#node#Require(host) abort
     let args = ['node']
 
-    if !empty($NVIM_NODE_HOST_DEBUG)
+    if provider#node#can_inspect() && !empty($NVIM_NODE_HOST_DEBUG)
```
`!empty($NVIM_NODE_HOST_DEBUG)` should be tested first, that avoids a process invocation.

justinmk commented on test/functional/provider/nodejs_spec.lua:
```diff
@@ -0,0 +1,68 @@
+
+describe('nodejs host', function()
+  it('works', function()
+    local fname = 'hello.js'
```
Test files always begin with "Xtest".

    local fname = 'Xtest-hello.js'

This helps things like `make clean`.

justinmk commented on test/functional/provider/nodejs_spec.lua:
```diff
@@ -0,0 +1,68 @@
+    os.remove(fname)
+  end)
+  it('plugin works', function()
+    local fname = 'hello-plugin.js'
```
local fname = 'Xtest-provider-nodejs-hello-plugin.js'

justinmk commented on test/functional/provider/nodejs_spec.lua:
```diff
@@ -0,0 +1,68 @@
+      nvim.command('call jobstop(g:job_id)');
+    ]])
+    command('let g:job_id = jobstart(["node", "'..fname..'"])')
+    sleep(5000)
```
Is this still needed? Can `helpers.retry()` be used instead?

justinmk commented on test/functional/provider/nodejs_spec.lua:
```diff
@@ -0,0 +1,68 @@
+    command('let g:job_id = jobstart(["node", "'..fname..'"])')
+    sleep(5000)
+    eq('hello', eval('g:job_out'))
+    os.remove(fname)
```
should go in `after_each()`

janlazo replied:
Perhaps teardown is better to guarantee deletion for all temporary test files.

justinmk commented on test/functional/provider/nodejs_spec.lua:
```diff
@@ -0,0 +1,68 @@
+    command('let g:job_id = jobstart(["node", "'..fname..'"])')
+    sleep(5000)
+    eq('hello-plugin', eval('g:job_out'))
+    os.remove(fname)
```
should go in `after_each()`

janlazo committed:
provider: check $NVIM_NODE_HOST_DEBUG first for --inspect-brk

janlazo committed:
health: improve nodejs warnings

janlazo committed:
test: rename nodejs temporary test files

janlazo committed:
test: speedup tests for nodejs host

- sleep for 500ms and then retry checking the output until 4.5s has passed
- guarantee deletion of temporary test files with busted's teardown()

janlazo commented on test/functional/provider/nodejs_spec.lua:
```diff
@@ -61,8 +66,7 @@ describe('nodejs host', function()
-    eq('hello-plugin', eval('g:job_out'))
-    os.remove(fname)
+    sleep(500)
+    retry(nil, 4500, function() eq('hello-plugin', eval('g:job_out')) end)
```
https://ci.appveyor.com/project/neovim/neovim/build/17222/job/xsuglue2xs24b8rs

```
[00:07:47] [ RUN      ] nodejs can inspect: 156.00 ms OK
[00:07:48] [ RUN      ] nodejs host works: 671.00 ms OK
[00:07:49] [ RUN      ] nodejs host plugin works: 532.00 ms OK
```

I guess `sleep(500)` isn't necessary after all. Each test should be done within 1 second.

janlazo committed:
Revert "host.vim: don't normalize slashes for nodejs rplugins"

This reverts commit 84a933fd671cf563e8022261a63917840b692d06.

This does nothing.
I forgot to set `NODE_PATH` on my system so I couldn't register any nodejs rplugins.

janlazo committed:
health: $NVIM_NODE_HOST_DEBUG is not supported on Windows

janlazo committed:
test: retry up to 1s for nodejs host tests to succeed

justinmk commented on runtime/autoload/health/provider.vim:
```diff
@@ -502,7 +502,14 @@ function! s:check_node() abort
-  call health#report_info('Node: '. s:system('node -v'))
+  let node_v = get(split(s:system('node -v'), "\n"), 0, '')
+  call health#report_info('Node: '. node_v)
+  if !s:shell_error && s:version_cmp(node_v[1:], '6.0.0') < 0
```
What if there is a s:shell_error?

janlazo replied:
I assume that `node -v` failed (how?) so I can't rely on its output or its behaviour when running `neovim-node-host`. 
I'd rather (add a warning and) return early here but the existing code continues on.

`provider#node#can_inspect` handles this case by returning early.

janlazo committed:
provider: use get() to avoid VimL list error

janlazo commented:
What's the expected behaviour of `:UpdateRemotePlugins` if python, ruby, node, etc. are not in `PATH`?
I get errors for node provider if `node` is not in `PATH` because it tries to run the nodejs host.
What about python and ruby?

janlazo commented:
Having second thoughts about `NVIM_NODE_HOST_DEBUG` because `--inspect-brk` is "experimental" to begin with.

```
PS> node --inspect-brk
Debugger listening on port 9229.
Warning: This is an experimental feature and could change at any time.
To start debugging, open the following URL in Chrome:
    chrome-devtools://devtools/remote/serve_file/@60cd6e859b9f557d2312f5bf532f6aec5f284980/inspector.html?experiments=true&v8only=true&ws=127.0.0.1:9229/912259bd-363b-48ef-9e9d-1bcaae733407
```

janlazo committed:
test: try nodejs rplugin on Travis

janlazo committed:
Revert "test: try nodejs rplugin on Travis"

This reverts commit 2de4a0260b659567cf8b36fbc8ebaa1e385ac0be.

This doesn't work. Leave it to the user since rplugins for python and ruby
aren't tested here anyway.

justinmk commented on runtime/autoload/health/provider.vim:
```diff
@@ -502,7 +502,14 @@ function! s:check_node() abort
+    call health#report_warn('nodejs host does not support '.node_v)
+  endif
+  if has('win32') || !provider#node#can_inspect()
+    call health#report_warn('nodejs on this system does not support --inspect-brk or this system is Windows so $NVIM_NODE_HOST_DEBUG is ignored.')
```
The message is getting too long, there's not much reason to use the same message when we can trivially check if we're on Windows.

```vim
if has('win32')
  call health#report_warn('nodejs on Windows does not support --inspect-brk. $NVIM_NODE_HOST_DEBUG will be ignored.')
elseif !provider#node#can_inspect()
  call health#report_warn('nodejs on this system does not support --inspect-brk. $NVIM_NODE_HOST_DEBUG will be ignored.')
endif
```

janlazo replied:
As long as the version requirement is reached, `--inspect-brk` works on Windows.
The problem is that you cannot pass this to `node` through the shims.

```sh
# Ignore the shims and run cli.js directly
node   $PREV  "$(npm root -g)/neovim/bin/cli.js"   $NEXT
```

Without calling `node` directly, the user can't set `$PREV`.

justinmk commented:
> Having second thoughts about NVIM_NODE_HOST_DEBUG because --inspect-brk is "experimental"

It's ok. We can always remove it later. The work has been done, it's always nice to have extra knowledge in the health checker.

janlazo committed:
provider: run cli.js of nodejs host directly

npm shims are useless because the user cannot set node to debug mode
via --inspect-brk. This is problematic on Windows which use batchfiles and shell scripts
to compensate for not supporting shebang.

The patch uses `npm root -g` to get the absolute path of the global npm modules.
If that fails, then the user did not install neovim npm package globally.
Use that absolute path to find `neovim/bin/cli.js`, which is what the npm shim
actually runs with node. glob() is for a simple file check in case bin/ is removed
because the npm shims are ignored now.

justinmk commented:
travis failure: not related to this PR, please ignore. ASAN complains because `loop_deferred_event()` wasn't called before nvim exited, perhaps after `nvim +q` or something quick like that.


```
=================================================================
==21832==ERROR: LeakSanitizer: detected memory leaks

Direct leak of 56 byte(s) in 1 object(s) allocated from:
    #0 0x511b26 in malloc (/home/travis/build/neovim/neovim/build/bin/nvim+0x511b26)
    #1 0x1009a84 in try_malloc /home/travis/build/neovim/neovim/src/nvim/memory.c:87:15
    #2 0x1009c44 in xmalloc /home/travis/build/neovim/neovim/src/nvim/memory.c:121:15
    #3 0xaa8c36 in loop_schedule_deferred /home/travis/build/neovim/neovim/src/nvim/event/loop.c:89:19
    #4 0x190856a in tui_main /home/travis/build/neovim/neovim/src/nvim/tui/tui.c:367:5
    #5 0x1963d61 in ui_thread_run /home/travis/build/neovim/neovim/src/nvim/ui_bridge.c:106:3
    #6 0x2b5d4190d183 in start_thread /build/eglibc-SvCtMH/eglibc-2.19/nptl/pthread_create.c:312
```

justinmk commented on runtime/autoload/health/provider.vim:
```diff
@@ -502,7 +502,14 @@ function! s:check_node() abort
+  let node_v = get(split(s:system('node -v'), "\n"), 0, '')
+  call health#report_info('Node: '. node_v)
+  if !s:shell_error && s:version_cmp(node_v[1:], '6.0.0') < 0
+    call health#report_warn('nodejs host does not support '.node_v)
```
Users are going to be confused, most don't know that "nodejs host" is something different than actual "nodejs". It's a bit of mess, but for user-facing messages we should say "Neovim node-client" (the literal repo name) or "Neovim nodejs host".

This is why `:help dev-api-client` now says very clearly:

```
Package Naming                                                                                                                                                  
API client packages should NOT be named something ambiguous like "neovim" or                                                                                    
"python-client".  Use "nvim" as a prefix/suffix to some other identifier                                                                                        
following ecosystem conventions. 
```

janlazo committed:
health: fix version check for Neovim node-client

janlazo committed:
health: fix node.js references for consistency

justinmk commented on test/functional/provider/nodejs_spec.lua:
```diff
@@ -0,0 +1,70 @@
+
+describe('nodejs', function()
+  it('can inspect', function()
+    eq(1, funcs['provider#node#can_inspect']())
```
This is going to fail on some developer systems, isn't it? It failed on mine...

janlazo replied:
Yes, because it's meant for CI only and it's common to have old node versions if installing from OS package manager, not nvm.

justinmk replied:
Ok, then why aren't the tests skipped on my system?

janlazo replied:
Since Travis and Appveyor use nodejs 6.12+, it's pointless to have it now, even as a sanity check.

justinmk commented on test/functional/provider/nodejs_spec.lua:
```diff
@@ -0,0 +1,70 @@
+local eval = helpers.eval
+local sleep = helpers.sleep
+local funcs = helpers.funcs
+local retry = helpers.retry
```
After doing:

    sudo npm install -g neovim

these tests fail on my system. I think `NODE_PATH` should be set in the tests, not in CI scripts only.

janlazo committed:
test: remove inspect test; set NODE_PATH in nodejs_spec.lua

provider#node#can_inspect will fail on some systems because it is common
to have old node versions in OS (any Linux OS that has LTS releases)
and CI (Travis, Appveyor).

NODE_PATH can be trivially set with VimL.
Build scripts don't have to set it for the nodejs tests to work.
NODE_PATH is optional to begin with and is used only as a workaround
for the neovim node.js host.

justinmk commented on test/functional/provider/nodejs_spec.lua:
```diff
@@ -20,12 +20,7 @@ end
-  it('can inspect', function()
-    eq(1, funcs['provider#node#can_inspect']())
-  end)
+  command([[let $NODE_PATH = get(split(system('npm root -g'), "\n"), 0, '')]])
```
Why not use systemlist() instead of split(system())?

janlazo replied:
CRLF on Windows.

janlazo commented:
@justinmk Does it work on your system now?

janlazo committed:
test: remove unused helpers to pass lint test

justinmk commented:
I have node 4.8.3, I'm guessing that's the problem. When I run the test plugin manually:

```
:!node Xtest-nodejs-hello.js                                                                                                                                    
                                                                                                                                                                
/home/vagrant/node_modules/neovim/lib/attach/attach.js:6                                                                                                        
function attach({ reader: _reader, writer: _writer, proc, socket, }) {                                                                                          
                ^                                                                                                                                               
                                                                                                                                                                
SyntaxError: Unexpected token {                                                                                                                                 
    at exports.runInThisContext (vm.js:53:16)                                                                                                                   
    at Module._compile (module.js:373:25)                                                                                                                       
    at Object.Module._extensions..js (module.js:416:10)                                                                                                         
    at Module.load (module.js:343:32)                                                                                                                           
    at Function.Module._load (module.js:300:12)                                                                                                                 
    at Module.require (module.js:353:17)                                                                                                                        
    at require (internal/module.js:12:17)                                                                                                                       
    at Object.<anonymous> (/home/vagrant/node_modules/neovim/lib/attach.js:3:16)                                                                                
    at Module._compile (module.js:409:26)                                                                                                                       
    at Object.Module._extensions..js (module.js:416:10)                                                                                                         
shell returned 1    
```

justinmk commented:
I added a version check in `provider#node#Detect()`. Will merge shortly.

janlazo commented:
@justinmk  Sure. I was about to do the same thing. Thanks.

justinmk commented:
Merged in https://github.com/neovim/neovim/pull/7734 . 

Thanks @janlazo , this definitely exposed some rough edges and answered questions that would have led to bug reports.

justinmk commented:
@janlazo travis log mentions these nodejs versions are pre-installed:

```
Pre-installed Node.js versions[0m
v4.8.6
v6.12.0
v6.12.1
v8.9
v8.9.1
```

Why don't we use one of those instead of using nvm.sh? Would be good to save time in the build where possible.

janlazo commented:
Which part of the build? Switching node versions or installation?
Travis uses nvm for node_js builds out of convenience. There is [n](https://github.com/tj/n) for globally installed node versions but it must be installed separately with `npm install -g n`.
